src/Microsoft.AspNet.Mvc.DataAnnotations/CompareAttributeAdapter.cs

```diff
@@ -28,12 +28,17 @@ public CompareAttributeAdapter(CompareAttribute attribute, IStringLocalizer stri
                 throw new ArgumentNullException(nameof(context));
             }
 
-            var errorMessage = ((CompareAttributeWrapper)Attribute).FormatErrorMessage(context);
+            var errorMessage = GetErrorMessage(context.ModelMetadata, context.MetadataProvider);
             var clientRule = new ModelClientValidationEqualToRule(errorMessage,
                                                             FormatPropertyForClientValidation(Attribute.OtherProperty));
             return new[] { clientRule };
         }
 
+        public string GetErrorMessage(ModelMetadata metadata, IModelMetadataProvider metdataProvider)
```

Comment:
Is this the only bit stopping you from exposing the default `protected` `GetErrorMessage` as `virtual`?